### ISSUE TITLE
Sophisticated content-id validation

### ISSUE THREAD

**Issue by snazy** [projectnessie/nessie#3832]
Nessie should perform a couple more checks and some checks in a different way:

* every `Put` operation for a _new_ content key must have an _empty_/no `expectedContent`
* every `Put` operation for an _existing_ content key must have a _non-empty_ `expectedContent`
* verify that the content id of the `content` in a `Put` operation is equal to the content id of a _non-empty_ `expectedContent`

Depending on the content type, Nessie should perform this check as well:
* for a non-empty `expectedContent`, verify that the content id already exists

Also move checks from `PersistVersionStore.commit()` to `AbstractDatabaseAdapter`

**Comment by dalaro** [comment]
I'm assuming this issue describes checks on `Put` operations within commits, otherwise my comment here is probably way off the mark.

I think some or all of this may have since been implemented in [`PersistVersionStore#commit`](https://github.com/projectnessie/nessie/blob/main/versioned/persist/store/src/main/java/org/projectnessie/versioned/persist/store/PersistVersionStore.java#L119-L154) by 1654cf5fe51d969fabe6c071d1a080856240e744.

While writing some tests to exercise these checks in [an experimental branch](https://github.com/dalaro/nessie/tree/put-content-id-checks) in my fork, I found at least one case where I was unsure about the target behavior.  When a content key already exists, a new commit containing a `Put` operation on that key can specify an arbitrary value for both the expected content-id and the content-id to be written by the op.  As long as this arbitrary value is specified exactly the same as both expectation and to-write, it need not match the content-id of whatever's currently resident at the key.  This is a tricky case for me to articulate, so maybe [these fifteen lines](https://github.com/dalaro/nessie/blob/put-content-id-checks/servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractRestInvalidWithHttp.java#L597-L611) better convey it:

```java
    /*
     * Creates an Operations singleton.  The sole element is a Put.
     * The expected and to-be-written content IDs are both "fake id".
     * This intentionally mismatches the server-assigned UUID of the existing content at this key.
     */
    Operations putWithMatchingFakeContentIds =
        getPutTableOpSingleton(
            "fake id", "fake id", contentKey, "put with matching fake content ids");
    // TODO should this fail?
    getHttpClient()
        .newRequest()
        .path("trees/branch/{branchName}/commit")
        .resolveTemplate("branchName", branchName)
        .queryParam("expectedHash", getHeadHash(branchName))
        .post(putWithMatchingFakeContentIds);
```

On the one hand, allowing this makes intuitive sense, since adding a check in this case seems to imply costly read-before-write on every committed `Put`, unless I'm ignoring a cache somewhere.  On the other hand, I couldn't tell whether the final bullet point (the content-type-dependent one) was describing this scenario.

**Comment by snazy** [comment]
Regarding that particular case w/ that `fake-id`, it should fail, because the content-id's different from the existing one.

**Comment by dalaro** [comment]
> Regarding that particular case w/ that `fake-id`, it should fail, because the content-id's different from the existing one.

Thank you.  I have three follow-up questions:

* Looking at `DatabaseAdapter`, some of its non-transactional implementations, and https://projectnessie.org/develop/kernel/, I'm not sure I see a cheap way to do this.  I think it can be done with `DatabaseAdapter#keys`, but that's a heavy read-before-write, potentially fetching up to 20 commits until `CommitLogEntry#hasKeySummary` is true as it reconstructs the net key-set.  I've [prototyped](https://github.com/dalaro/nessie/blob/put-content-id-checks/versioned/persist/store/src/main/java/org/projectnessie/versioned/persist/store/PersistVersionStore.java#L175-L201) this use of `DA#keys` already (not review-ready).  It feels like I'm on the wrong track here from an efficiency standpoint.  Am I missing an index structure that I should use to answer this, maybe avoiding `DA#keys`?
* What content types should be subject to this check?
* When I turned on this check, I think I found a latent bug in `AbstractResteasyTest`.  If you have a moment, could you verify that this [tentative fix](https://github.com/dalaro/nessie/commit/ffd742aea51f806b7ddfb98f83ed90376f0ccf86#diff-af332f84ec03c03dc94fcadedc40c7d0dc423887e30045f2eaf7c61a75849ddf) makes sense in context, where `xxx.test` is updated multiple times?  Without this change, the ensuing commit would throw a conflict.

**Comment by snazy** [comment]
Can you open a (draft) PR for the changes? That makes it much easier to comment on code.

Regarding the check: yes, that read-before-write is necessary. The check's placed in the correct place - but it should run for all operations and:
* run from `CommitParams.validator()` (a callback from the database-adapter) - the callback ensures that you're checking against the latest version.
* from there: get all keys using the key-filter-predicate - there must be no content-key for put-operations w/o an expected-content - for put-operations w/ an expected-content, the content-id must match

We can of course update the functionality in `(Abstract)DatabaseAdapter` to become more efficient for this use case - aka let the `keys()` function take a `Collection<Key> keys` as `values()` does.

Re the "tentative fix": IIUC w/o the tentative fix the commit fails, because the new check triggers. If that's the case, then the fix looks correct to me. But it could also use the value from `updates[10]` (or memoize the value for that one).

**Comment by dalaro** [comment]
> Can you open a (draft) PR for the changes? That makes it much easier to comment on code.

Will do; I'd been delaying since I wanted to clean up my branch a bit first.  Thanks for the answers to each of my questions, and the additional pointers.  I appreciate it.

> IIUC w/o the tentative fix the commit fails, because the new check triggers.

That's correct.  The new check plus an old test fails on the ensuing commit, [specifically L194 @ dd8a041d](https://github.com/projectnessie/nessie/blob/dd8a041d1699091b7e05a20e631d9254164038a1/servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/AbstractResteasyTest.java#L194).  The client-side `AssertionError` message is `Expected status code <200> but was <400>,` and logged stdout from the server shows a response in this form:

```
HTTP/1.1 400 Bad Request
Content-Type: application/json
content-length: 281

{
    "status": 400,
    "reason": "Bad Request",
    "message": "Conflict between expected content-id '0f158b34-93dc-4d4d-aea9-87e194ddad3a' and actual content-id '8a35d258-3f2b-4007-b405-7eb8ed09ecd5' for key 'xxx.test'",
    "errorCode": "BAD_REQUEST",
    "serverStackTrace": null
}
```


Also, I haven't forgotten about this:

> Also move checks from `PersistVersionStore.commit()` to `AbstractDatabaseAdapter`

but I'm leaving it for the end right now.

**Comment by snazy** [comment]
Fixed with #5641
